### PR TITLE
Add multi-stage full build and test

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
This creates jobs to build all of IREE on one machine and then test on
another, passing the build directory around. This same build directory
can be used in other jobs also (e.g. for GPU testing, which I'm punting
on for now).

I ran into issues using the GitHub action artifacts for this. I knew
we'd get suboptimal transfer speeds since it's hosted on Azure and
we'd probably want another solution eventually, but they were much
worse than I expected. So this uses a GCS bucket for the transfer
instead.

We don't want presubmit jobs to be able to affect builds from other
commits, so that's a bit tricky. To get around this, I have separate
buckets for presubmit and postsubmit. The runners only have "create"
and "read" permissions, not "delete" permissions, which means they
cannot remove or overwrite existing artifacts. They upload to a file
including the commit name in its path. If a file is already there, the
upload (and therefore the job) would fail. I think this avoids the
possibility of one build affecting another, even if someone were trying
to do so intentionally. If we move to avoiding rerunning a job if the
artifact for it already exists, we'll need to revisit this.

Part of https://github.com/iree-org/iree/issues/9881

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
Hmm, maybe we'll do GPU in a follow-up...

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -69,6 +71,105 @@ jobs:
...
+      - name: "Uploading build dir archive"
+        id: upload
+        env:
+          GCS_ARTIFACT: ${{ env.BUCKET }}/${{ github.sha }}/${{ steps.archive.outputs.build-dir-archive }}
+          BUILD_DIR_ARCHIVE: ${{ steps.archive.outputs.build-dir-archive }}
+        run: |
+          gcloud alpha storage cp "${BUILD_DIR_ARCHIVE}" "${GCS_ARTIFACT}"
+          echo "::set-output name=gcs-artifact::${GCS_ARTIFACT}"
+
+
```
Comment (pzread) [inline .github/workflows/ci.yml:134]: nit: Extra blank line?

File: build_tools/cmake/build_all.sh [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#!/bin/bash
+
+# Copyright 2019 The IREE Authors
```
Comment (pzread) [inline build_tools/cmake/build_all.sh:3]: nit: We can use 2022 for new file?

**Review by pzread — APPROVED** [review]

File: build_tools/cmake/build_all.sh (hunk 2 above)
Comment (GMNGeoffrey) [reply build_tools/cmake/build_all.sh:3]: I forked it from an old file. Git doesn't always follow the renames. And we're supposed to preserve the oldest applicable copyright year. I really wish the Apache license hadn't required a copyright year... :-/

File: .github/workflows/ci.yml (hunk 1 above)
Comment (GMNGeoffrey) [reply .github/workflows/ci.yml]: Done

File: build_tools/cmake/build_all.sh (hunk 2 above)
Comment (marbre) [reply build_tools/cmake/build_all.sh:3]: Following here by accident: I think the Apache license only suggests to add a copyright year but I don‘t think there is a strict requirement.

Comment (GMNGeoffrey) [reply build_tools/cmake/build_all.sh:3]: I've talked with Google lawyers about this at unfortunate length. It was only through much complaining that I got our license header as short as it is :-(

**Comment by GMNGeoffrey** [comment]
Forgot to tag originally, but part of https://github.com/iree-org/iree/issues/9855

**Pull request merged**